c7n/resources/s3.py

```diff
@@ -783,6 +785,165 @@ def process_bucket(self, b):
+    bucket will match.
+
+    Conversely, if you specify a wildcard in a filter (e.g `*`), this will match all bucket
+    policy permissions.
+
+    Special considerations:
+
+    * Principals should be specified as "prefix:value" where prefix is one of AWS,
+      Federated, Service, etc.
+    * Partial wildcards (e.g. s3:Put*) cannot be used in filter definition.
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: s3-public-writable-policy
+                resource: s3
+                filters:
+                  - type: has-policy-permission
+                    actions:
+                      - s3:PutObject
+                      - s3:PutObjectAcl
+                      - s3:PutBucketPolicy
+                    principals:
+                      - 'AWS:*'
+
+            policies:
+              - name: s3-public-access-policy
+                resource: s3
+                filters:
+                  - type: has-policy-permission
+                    actions:
+                      - '*'
+                    principals:
+                      - '*'
+    """
+    schema = type_schema(
+        'has-policy-permission',
+        actions={'type': 'array', 'items': {'type': 'string'}},
+        principals={'type': 'array', 'items': {'type': 'string'}}
+    )
+
+    def process(self, buckets, event=None):
+        return list(filter(None, map(self.process_bucket, buckets)))
+
+    def process_bucket(self, b):
+        p = b.get('Policy')
+        if p is None:
+            return None
+        p = json.loads(p)
+
+        required_actions = list(self.data.get('actions', []))
+        for required_action in required_actions:
+            if ("*" in required_action and len(required_action) > 1):
+                raise ValueError(
+                    "partial wildcards in action filters are not supported: %s" %
+                    required_action)
+
+        required_principals = list(self.data.get('principals', []))
+        for required_principal in required_principals:
+            if ("*" in required_principal and len(required_principal) > 1 and
+                    not re.match('^[^:]+:\*$', required_principal)):
+                raise ValueError(
+                    "partial wildcards in principal filters are not supported: %s" %
+                    required_principal)
+
+        statements = p.get('Statement', [])
+        for statement in statements:
+            if statement.get('Effect', 'Allow').lower() == 'deny':
+                continue
+
+            matched_actions = self.__get_action_matches(statement, required_actions)
+            matched_principals = self.__get_principal_matches(statement, required_principals)
+
+            if ((len(required_actions) == 0 or matched_actions) and
+                    (len(required_principals) == 0 or matched_principals)):
+                return b
+
+        return None
+
+    def __get_action_matches(self, statement, required_actions):
```

Comment:
one underscore is fine to denote private, class mangling from double __ is typically gratuitous and just as easy to bypass.